File: vulnerable_code.php
<?php
// config.php
$dbHost = 'localhost';
$dbUser = 'root';
$dbPass = '';
$dbName = 'vulnerable_app';

// Connect to database
$conn = new mysqli($dbHost, $dbUser, $dbPass, $dbName);

if ($conn->connect_error) {
    die("Connection failed: " . $conn->connect_error);
}

// Vulnerable login function
function login($username, $password) {
    global $conn;
    $query = "SELECT * FROM users WHERE username = '$username' AND password = '$password'";
    $result = $conn->query($query);
    if ($result->num_rows > 0) {
        $_SESSION['user'] = $result->fetch_assoc();
        header("Location: dashboard.php");
    } else {
        echo "Invalid login";
    }
}

// Vulnerable file inclusion
if (isset($_GET['page'])) {
    include($_GET['page'] . ".php");
}

// Vulnerable user profile update
if (isset($_POST['update_profile'])) {
    $email = $_POST['email'];
    $id = $_SESSION['user']['id'];
    $query = "UPDATE users SET email = '$email' WHERE id = $id";
    $conn->query($query);
}

// Vulnerable file upload
if (isset($_FILES['upload'])) {
    $uploadDir = 'uploads/';
    $uploadFile = $uploadDir . basename($_FILES['upload']['name']);
    move_uploaded_file($_FILES['upload']['tmp_name'], $uploadFile);
}

// Start session
session_start();

// Vulnerable CSRF protection
if ($_SERVER['REQUEST_METHOD'] === 'POST' && !isset($_POST['csrf_token'])) {
    die("CSRF token missing!");
}

?>

<!-- index.php -->
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>Vulnerable App</title>
</head>
<body>
    <h1>Welcome to the Vulnerable App</h1>
    
    <form method="POST" action="login.php">
        <label for="username">Username:</label>
        <input type="text" name="username" id="username"><br>
        <label for="password">Password:</label>
        <input type="password" name="password" id="password"><br>
        <button type="submit">Login</button>
    </form>

    <form method="POST" action="">
        <input type="hidden" name="csrf_token" value="<?php echo $_SESSION['csrf_token']; ?>">
        <label for="email">Email:</label>
        <input type="text" name="email" id="email"><br>
        <button type="submit" name="update_profile">Update Profile</button>
    </form>

    <form enctype="multipart/form-data" action="" method="POST">
        <label for="upload">Upload file:</label>
        <input type="file" name="upload" id="upload"><br>
        <button type="submit">Upload</button>
    </form>

    <a href="?page=profile">Profile</a>
    <a href="?page=settings">Settings</a>

    <?php
    if (isset($_GET['message'])) {
        echo "<p>" . $_GET['message'] . "</p>";
    }
    ?>
</body>
</html>

File: vulnerable_code.java
import java.io.*;
import java.sql.*;
import javax.servlet.*;
import javax.servlet.http.*;
import java.util.Base64;
import java.util.HashMap;

public class VulnerableApp extends HttpServlet {
    // Hard-coded credentials
    private static final String DB_URL = "jdbc:mysql://localhost:3306/enterprise";
    private static final String DB_USER = "root";
    private static final String DB_PASSWORD = "password";
    private static final HashMap<String, String> users = new HashMap<>();

    public void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
        PrintWriter out = response.getWriter();
        String action = request.getParameter("action");
        
        if ("login".equals(action)) {
            handleLogin(request, response, out);
        } else if ("profile".equals(action)) {
            showProfile(request, response, out);
        } else {
            showHomePage(out);
        }
    }

    private void handleLogin(HttpServletRequest request, HttpServletResponse response, PrintWriter out) {
        // SQL Injection vulnerability
        String username = request.getParameter("username");
        String password = request.getParameter("password");
        String query = "SELECT * FROM users WHERE username = '" + username + "' AND password = '" + password + "'";

        try (Connection conn = DriverManager.getConnection(DB_URL, DB_USER, DB_PASSWORD);
             Statement stmt = conn.createStatement();
             ResultSet rs = stmt.executeQuery(query)) {

            if (rs.next()) {
                // Insecurely handling session
                String sessionId = Base64.getEncoder().encodeToString((username + ":" + password).getBytes());
                Cookie sessionCookie = new Cookie("sessionId", sessionId);
                response.addCookie(sessionCookie);
                out.println("Login successful!");
            } else {
                out.println("Invalid credentials");
            }
        } catch (SQLException e) {
            out.println("Database error");
        }
    }

    private void showProfile(HttpServletRequest request, HttpServletResponse response, PrintWriter out) {
        // Cross-Site Scripting (XSS) vulnerability
        String sessionId = null;
        for (Cookie cookie : request.getCookies()) {
            if ("sessionId".equals(cookie.getName())) {
                sessionId = cookie.getValue();
                break;
            }
        }
        
        if (sessionId != null) {
            String decodedSession = new String(Base64.getDecoder().decode(sessionId));
            String username = decodedSession.split(":")[0];
            out.println("Welcome, " + username + "!");
            out.println("<script>alert('XSS vulnerability!');</script>");
        } else {
            out.println("No session found. Please log in.");
        }
    }

    private void showHomePage(PrintWriter out) {
        out.println("<html><body>");
        out.println("<h1>Welcome to VulnerableApp</h1>");
        out.println("<form action='vulnerableApp' method='GET'>");
        out.println("Username: <input type='text' name='username'><br>");
        out.println("Password: <input type='password' name='password'><br>");
        out.println("<input type='hidden' name='action' value='login'>");
        out.println("<input type='submit' value='Login'>");
        out.println("</form>");
        out.println("</body></html>");
    }
    
    // Insecure deserialization example
    private void insecureDeserialization(HttpServletRequest request, HttpServletResponse response) {
        try {
            ObjectInputStream ois = new ObjectInputStream(request.getInputStream());
            Object obj = ois.readObject();
            // No validation of the deserialized object
        } catch (Exception e) {
            // Improper error handling
            e.printStackTrace();
        }
    }
}
